src/test/java/BaseTest.java

```diff
@@ -1,11 +1,88 @@
 import io.github.bonigarcia.wdm.WebDriverManager;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeDriver;
 import org.openqa.selenium.chrome.ChromeOptions;
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.Parameters;
+
+import java.time.Duration;
 
 public class BaseTest {
 
+    protected  WebDriver driver = null;
+
+    public WebDriverWait wait ;
+
+    public static Actions  actions = null ;
+
+    //public WebDriver driver;
+
     @BeforeSuite
     static void setupClass() {
         WebDriverManager.chromedriver().setup();
     }
-}
\ No newline at end of file
+
+
+    @BeforeMethod
+    @Parameters ("baseUrl")
+
+    public void launchBrowser(String baseUrl) {
+        ChromeOptions options = new ChromeOptions();
+        options.addArguments("--remote-allow-origins=*");
+        driver = new ChromeDriver(options);
+        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
+        driver.manage().window().maximize();
+        wait = new WebDriverWait(driver,Duration.ofSeconds(10));
+
+        actions = new Actions(driver);
+
+        navigateToPage(baseUrl);
+
+
+    }
+    @AfterMethod
+    public void closeBrowser (){
+        driver.quit();
+    }
+
+
+
+
+    public void clickLoginBtn() {
+        // WebElement loginBtn = driver.findElement(By.cssSelector("button[type='submit']"));
+        WebElement loginBtn = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("button[type='submit']")));
+        loginBtn.click();
+    }
+
+    public  void providePassword(String password) {
+        WebElement passwordField = wait .until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("input[type='password']")));
+        passwordField.clear();
+        passwordField.sendKeys("Password2026@");
+    }
+
+    public void provideEmail(String email) {
+        //WebElement emailField = driver.findElement(By.cssSelector("input[type='email']"));
+        WebElement emailField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("input[type='email']")));
+        emailField.clear();
+        emailField.sendKeys("tassadit.talbi@testpro.io");
+    }
+
+    public void navigateToPage(String baseUrl) {
+        //String url = "https://qa.koel.app/";
+        driver.get(baseUrl);
```

Comment:
**BaseTest** — Implements the necessary methods for work with webdriver.

The BaseTest class contains common setup and teardown methods, as well as any shared functionality or configurations required by multiple test classes.

**BaseTest**

The problem we are trying to solve here is the reuse of common behaviors across test classes avoiding coding duplication and centralizing these actions at one point.

**@BeforeSuite**	- The annotated method will be run before all tests in this suite have run.

**@BeforeMethod** -	The annotated method will be run before all the test methods in the current class have been run.

**@AfterMethod**	- The annotated method will be run after each test method.

_Example without a Base Test Class_

```
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.testng.annotations.*;

import java.time.Duration;

public class BaseTest {
    public static WebDriver driver = null;

    @BeforeSuite
    static void setupClass() {
        WebDriverManager.chromedriver().setup();
    }


    @BeforeMethod
    @Parameters({"BaseURL"})
    public void launchBrowser(String baseURL) {
        ChromeOptions optionsChrome = new ChromeOptions();
        optionsChrome.addArguments("--disable-notifications","--remote-allow-origins=*", "--incognito","--start-maximized");
        optionsChrome.setExperimentalOption("excludeSwitches", new String[]{"enable-automation"});
        driver = new ChromeDriver(optionsChrome);
        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
        driver.manage().window().maximize();
        driver.get(baseURL);

    }
    @AfterMethod
    public void closeBrowser() {
        driver.quit();
    }
}
```

![image](https://github.com/Test-Pro-Bootcamp/A57-QA-Automation-Framework/assets/104761482/1af46f9b-39dc-4a95-8f2d-d7d2adcbee07)